src/models/losses.py

```diff
@@ -4,6 +4,327 @@
+
+        vgg_encoder: The VGG encoder model.
+
+        real_imgs: The Real Images from the dataset. shape: (batch_size, 3, 256, 256)
+
+        device: The device to run the model on.
+
+        const_dict: The dictionary containing the constants.
+    """
+    total_error_g = 0
+    lambda3 = const_dict["lambda3"]
+
+    cond_logits = logits["fake"]["cond"]
+    cond_err_g = binary_cross_entropy(cond_logits, real_labels)
+
+    uncond_logits = logits["fake"]["uncond"]
+    uncond_err_g = binary_cross_entropy(uncond_logits, real_labels)
+
+    loss_g = (
+        cond_err_g + uncond_err_g
+    )  # add up the conditional and unconditional losses
+    total_error_g += loss_g
+
+    local_incept_feat, global_incept_feat = inception_encoder(fake_imgs)
+    l1_w, l2_w, l1_s, l2_s = damsm_loss(
+        local_incept_feat,
+        global_incept_feat,
+        words_emb,
+        sent_emb,
+        match_labels,
+        cap_lens,
+        class_ids,
+        device,
+        const_dict,
+    )  # DAMSM Loss from attnGAN.
+
+    loss_damsm = lambda3 * (l1_w + l2_w + l1_s + l2_s)
+
+    total_error_g += loss_damsm
+
+    vgg_real = vgg_encoder(real_imgs)  # shape: (batch, 128, 128, 128)
+    vgg_fake = vgg_encoder(fake_imgs)  # shape: (batch, 128, 128, 128)
+
+    loss_per = mean_squared_error(vgg_real, vgg_fake)  # perceptual loss
+
+    total_error_g += loss_per / 2.0
+    return total_error_g
+
+
+def damsm_loss(
+    local_incept_feat: torch.Tensor,
+    global_incept_feat: torch.Tensor,
+    words_emb: torch.Tensor,
+    sent_emb: torch.Tensor,
+    match_labels: torch.Tensor,
+    cap_lens: torch.Tensor,
+    class_ids: torch.Tensor,
+    device: Any,
+    const_dict: dict[Any, Any],
+) -> Any:
+    """Calculate the DAMSM loss from the attnGAN paper.
+
+    Args:
+        local_incept_feat: The local inception features. [shape: (batch, D, 17, 17)]
+
+        global_incept_feat: The global inception features. [shape: (batch, D)]
+
+        words_emb: The embeddings for all the words in the captions.
+
+        shape: (batch, D, max_caption_length)
+
+        sent_emb: The embeddings for the sentences. shape: (batch_size, D)
+
+        match_labels: Tensor of shape: (batch_size, 1).
+        This is of the form torch.tensor([0, 1, 2, ..., batch-1])
+
+        cap_lens: The length of the 'actual' captions in the batch [without padding]
+        shape: (batch_size, 1)
+
+        class_ids: The class ids for the instance. shape: (batch, 1)
+
+        device: The device to run the model on.
+
+        const_dict: The dictionary containing the constants.
+    """
+    batch_size = match_labels.size(0)
+    masks = (
+        []
+    )  # Mask mis-match samples, that come from the same class as the real sample ###
+
+    match_scores = []
+    cap_len_list = cap_lens.data.tolist()
+    gamma1 = const_dict["gamma1"]
+    gamma2 = const_dict["gamma2"]
+    gamma3 = const_dict["gamma3"]
+
+    for i in range(batch_size):
```

Comment:
I understand that it's easier to make it a loop and process each sample individually however I think it's beneficial for performance to try to use the vectorized operations on batch tensors

slow:
```
result = []
for i in range(t.size(0)):
    result.append(op(t[i]))
```

fast:
```
result = t.op()
```

Can you give it a thought?

Reply:
agree with you, but would prefer to do it later. Feel like not a top priority at the moment [and also bit busy with other projects and work]. Let's get started with results/experiments for now.

Reply:
> would prefer to do it later

It will save time if next time you do things that are efficient as early as possible. Because you may end up leaving an inefficient code because _"it works right?"_ , then you settle, and then you never fix it. 


For now resolved